pkg/daemon/ceph/client/upgrade.go

```diff
@@ -104,6 +157,197 @@ func EnableNautilusOSD(context *clusterd.Context) error {
 		return fmt.Errorf("failed to disallow pre-nautilus osds and enable all new nautilus-only functionality: %+v", err)
 	}
 	logger.Infof("successfully disallowed pre-nautilus osds and enabled all new nautilus-only functionality")
+	return nil
+}
+
+// OkToStop determines if it's ok to stop an upgrade
+func OkToStop(context *clusterd.Context, namespace, deployment, clusterName string, cephVersion cephver.CephVersion) error {
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		logger.Warningf("%+v", err)
+		return nil
+	}
+
+	// The ok-to-stop command for mon and mds landed on 14.2.1
+	// so we return nil if that Ceph version is not satisfied
+	if !cephVersion.IsAtLeast(cephver.CephVersion{Major: 14, Minor: 2, Extra: 1}) {
+		if daemonName != "osd" {
+			return nil
+		}
+	}
+
+	versions, err := GetCephVersions(context, clusterName)
+	if err != nil {
+		return fmt.Errorf("failed to get ceph daemons versions. %+v", err)
+	}
+
+	switch daemonName {
+	// Trying to handle the case where a **single** mon is deployed and an upgrade is called
+	case "mon":
+		// if len(versions.Mon) > 1, we have different Ceph versions for some monitor(s)
+		// this is fine running the upgrade we can run the upgrade checks
+		if len(versions.Mon) == 1 {
+			// now trying to parse and find how many mons are presents
+			// if we have less than 3 mons we skip the check and do best-effort
+			// we do less than 3 because during the initial bootstrap the mon sequence is updated too
+			// so running running the check on 2/3 mon fails
+			// versions.Mon looks like this map[ceph version 15.0.0-12-g6c8fb92 (6c8fb920cb1d862f36ee852ed849a15f9a50bd68) octopus (dev):1]
+			// now looping over a single element since we can't address the key directly (we don't know its name)
+			for _, monCount := range versions.Mon {
+				if monCount < 3 {
+					logger.Infof("the cluster has less than 3 monitors, not performing upgrade check, running in best-effort")
+					return nil
+				}
+			}
+		}
+	// Trying to handle the case where a **single** osd is deployed and an upgrade is called
+	case "osd":
+		// if len(versions.Osd) > 1, we have different Ceph versions for some osd(s)
+		// this is fine running the upgrade we can run the upgrade checks
+		if len(versions.Osd) == 1 {
+			// now trying to parse and find how many osds are presents
+			// if we have less than 3 osds we skip the check and do best-effort
```

Comment:
I understand why 3 for mons, but why 3 for OSDs? Couldn't this be larger than 3 depending on how failure domain is specified?